.golangci.yml

```diff
@@ -1,4 +1,15 @@
 linters-settings:
+  gocritic:
+    enabled-tags:
+      - diagnostic
+      - experimental
```

Comment:
Do we want to rely on experimental gocritic components in general? Some of them might be worth enabling piece-meal, but I’m not sure enabling the whole tag is a good idea.

Reply:
I'm not sure, good question. I think the experimental tag caught a lot of real things to improve relative to the few false positives/rules we don't like. Running with only the experimental tag against master, I get:

```
pkg/util/util.go:122:2: commentFormatting: put a space between `//` and comment text (gocritic)
	// length is 11
	^
pkg/util/util.go:241:9: elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
	} else {
	       ^
pkg/util/util.go:242:6: sloppyReassign: re-assignment to `err` can be replaced with `err := ipt.Insert(table, chain, 1, ruleSpec...)` (gocritic)
		if err = ipt.Insert(table, chain, 1, ruleSpec...); err != nil {
		   ^
pkg/util/util.go:101:68: wrapperFunc: use strings.ReplaceAll method in `strings.Replace(privateIP, ".", "-", -1)` (gocritic)
			CableName:     fmt.Sprintf("submariner-cable-%s-%s", clusterID, strings.Replace(privateIP, ".", "-", -1)),
			                                                                ^
pkg/cable/strongswan/strongswan.go:418:5: sloppyReassign: re-assignment to `err` can be replaced with `err := i.renderCharonConfigTemplate(f)` (gocritic)
	if err = i.renderCharonConfigTemplate(f); err != nil {
	   ^
pkg/cable/wireguard/driver.go:44:2: commentFormatting: put a space between `//` and comment text (gocritic)
	//TODO use submariner prefix
	^
pkg/cable/wireguard/driver.go:230:27: commentFormatting: put a space between `//` and comment text (gocritic)
			Port: w.spec.NATTPort, //TODO move port to endpoint spec
			                       ^
pkg/cable/libreswan/libreswan.go:323:7: badCond: `j == 0 && j == len(connections)-1` condition is suspicious (gocritic)
			if j == 0 && j == len(connections)-1 {
			   ^
pkg/cableengine/cableengine.go:83:5: sloppyReassign: re-assignment to `err` can be replaced with `err := i.driver.Init()` (gocritic)
	if err = i.driver.Init(); err != nil {
	   ^
pkg/cableengine/fake/cableengine.go:61:2: unnecessaryDefer: defer e.Unlock() is placed just before return (gocritic)
	defer e.Unlock()
	^
pkg/cableengine/syncer/syncer.go:62:2: ifElseChain: rewrite if-else to switch statement (gocritic)
	if errors.IsNotFound(err) {
	^
pkg/controllers/datastoresyncer/datastoresyncer.go:85:3: nestingReduce: invert if cond, replace body with `continue`, move old body after the statement (gocritic)
		if !util.CompareEndpointSpec(endpoints[i].Spec, d.localEndpoint.Spec) {
		^
pkg/routeagent/controllers/route/route.go:138:38: captLocal: `ClusterCidr' should not be capitalized (gocritic)
func NewController(clusterID string, ClusterCidr []string, ServiceCidr []string, objectNamespace string,
                                     ^
pkg/routeagent/controllers/route/route.go:799:10: elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
		} else {
		       ^
pkg/routeagent/controllers/route/route.go:901:11: elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
			} else {
			       ^
pkg/controllers/datastoresyncer/datastoresyncer_suite_test.go:15:2: dupImport: package is imported 2 times under different aliases on lines 15 and 16 (gocritic)
	clientsetv1 "github.com/submariner-io/submariner/pkg/client/clientset/versioned/typed/submariner.io/v1"
	^
pkg/controllers/datastoresyncer/datastoresyncer_suite_test.go:16:2: dupImport: package is imported 2 times under different aliases on lines 15 and 16 (gocritic)
	submarinerClientset "github.com/submariner-io/submariner/pkg/client/clientset/versioned/typed/submariner.io/v1"
	^
test/e2e/framework/framework.go:226:1: unnamedResult: consider giving a name to these results (gocritic)
func (f *Framework) GetGatewayInformer(cluster framework.ClusterIndex) (cache.SharedIndexInformer, chan struct{}) {
^
test/e2e/framework/submariner_resources.go:23:1: unnamedResult: consider giving a name to these results (gocritic)
func NoopCheckEndpoint(endpoint *submarinerv1.Endpoint) (bool, string, error) {
^
main.go:107:18: appendAssign: append result not assigned to the same slice (gocritic)
		localSubnets = append(submSpec.ServiceCidr, submSpec.ClusterCidr...)
		               ^
main.go:110:5: emptyStringTest: replace `len(submSpec.CableDriver) == 0` with `submSpec.CableDriver == ""` (gocritic)
	if len(submSpec.CableDriver) == 0 {
	   ^
pkg/globalnet/controllers/ipam/globalnet.go:9:2: dupImport: package is imported 2 times under different aliases on lines 9 and 10 (gocritic)
	k8sv1 "k8s.io/api/core/v1"
	^
pkg/globalnet/controllers/ipam/globalnet.go:10:2: dupImport: package is imported 2 times under different aliases on lines 9 and 10 (gocritic)
	v1 "k8s.io/api/core/v1"
	^
```

Reply:
Removed the experimental tag in the latest. :+1: